### ISSUE TITLE
Question: Best practices for handling customxpathfunction module in the project using CI/CD for BWCE app

### ISSUE THREAD

**Issue by bibhuduttaGitHub** [TIBCOSoftware/bw6-plugin-maven#505]
Maven Plugin version: 2.7.0

Maven version:3.4.3

Product : Either of TIBCO ActiveMatrix Businessworks 6 OR TIBCO Businessworks Container edition

Product version: BWCE 2.5.4

Component: Either of Maven Eclipse Plugin or Maven build plugin

**Steps to reproduce the issue:**
Team,
We are having a bunch of projects where we have migrated from BW 5.x to BWCE. And all of the  projects were having java custom functions so, they all are migrated to projects having customxpathfunction module in BWCE 2.5.4.
As the customxpathfunction modules are having multiple jars bundled in a jar.
We are looking for the best practices on how to handle such scenarios in a CI/CD workflow.

Currently , we are following the below approach for this after trying with few combinations to make it run:
1. Mavenising the Project (including customxpathfunction module). 
2. Commenting the customxpathfunction module from the .parent  POM.
3. Getting the customxpathfunction.jar(after renaming to Module's name_Customxpathfunction.jar) from the xpathfunction module and deploying to JFrog for external dependency management.
4. Adding the JFrog's dependency tags(obtained from Step-3) to the BWModule's POM.
5. Removing the Customxpathfunction module from the .application module's dependency.(done to avoid impaired state of the app post deployment).

Is there a more cleaner approach to handle the customxpathfunction modules in a Project? Which should work seamlessly like the projects having no customxpathfunction module. I would suggest if someone creates a wiki on how to achieve this.

We are using following versions:
BWCE: 2.5.4
Maven plugin: 2.7.0
Maven: 3.4.3
---------------------------------------------------

**Additional environment details if any:**

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by bibhuduttaGitHub** [comment]
Can i expect an update here..

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub, Please refer wiki page - https://github.com/TIBCOSoftware/bw6-plugin-maven/wiki/Custom-XPath-Function-with-Maven

**Comment by bibhuduttaGitHub** [comment]
![customxpath](https://user-images.githubusercontent.com/16704671/95602958-46c89780-0a73-11eb-88c5-58f4a393eaaf.PNG)
HI Vishal, 
Thanks for the wiki. We have a slightly different usecase here.
ref the screenshot: After migration, we are getting customXPathFunction.jar for each customXPathFunction module. And also,as you can see the name of the jars are always same (customXPathFunction.jar) irrespective of the module name. We need to find a way to rename the jar as we are keeping those jars in Artifactory for all the Projects and pulling them during mvn build through dependency management. Renaming of customXPathFunction.jar is needed  to ensure each Project has its own jar and they do not coincide with each other. If you can let us know how we can achieve this .

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub As discussed earlier, Please decide whether to keep the jars in lib folder or add it as maven dependency in pom.xml. I can see `EDWAdapter_customXPathFunction.jar` present in lib folder, added in classpath and it is also added as pom dependency.
![image](https://user-images.githubusercontent.com/29373785/95943238-e33bc280-0e02-11eb-9499-b83838ce849e.png)

The `EDWAdapter_customXPathFunction.jar` present in lib folder still has packaging issues. Please refer -
![image](https://user-images.githubusercontent.com/29373785/95945468-28162800-0e08-11eb-8c82-a7b09806295c.png)

To summerize, Please run below checks for each application before packaging using maven command -

- Review all dependencies required for the project. If it is custom xpath jar migrated from BW5.x, check jar packaging by running `jar tf filename.jar` command.
- Publish jar fixed in above step on repository or replace it in lib folder.
- follow the steps at wiki - https://github.com/TIBCOSoftware/bw6-plugin-maven/wiki/Custom-XPath-Function-with-Maven
- Execute `mvn clean package`

Thanks.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : MANIFEST file in 2nd entry in the jar is acceptable, right? In the last update, it was mentioned that it should be a first/second entry in the jar.

**Comment by vpawar-ai** [comment]
@bibhuduttaGitHub The Jar has only 3 files. :) As per links shared in #502, The discussion on JDK behavior is beyond the scope of this forum. after fixing the jar, it looks like this -
![image](https://user-images.githubusercontent.com/29373785/95948202-c062db80-0e0d-11eb-931d-26d5159825e4.png)

**Comment by bibhuduttaGitHub** [comment]
HI @vpawar-ai : EDWAdapter_customXPathFunction.jar -- gets generated out of the MIgration utility in Studio. If this jar is not as per the jdk packaging, it needs to be raised to TIBCO as a bug. However i have fixed the customxpathfunction  jar packaging. Now i can see the customxpathfunction's build is getting success and its failing with the BW module now. I have bundled all the jars the module has dependency to the case. I don't see any jar packaging issue with those jars. Have a look. Still i see null pointer exception.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : Did you check the latest debug log attached to the case?

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub, Yes, it looks similar. For migration utility issue, Please continue the discussion on the TIBCO support case. I have notified the team and they will continue to help on the issue. Thanks.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : BUt after fixing customxpathfunction.jar packaging, the customxpathfunction module got success. Now the null pointer is coming during build of BWModule. what could be the reason?

**Comment by vpawar-ai** [comment]
@bibhuduttaGitHub, I installed the jars from edw_adapter.7z into local m2 repo and executed `mvn clean package`. It worked fine. Attaching logs for your reference.
[maven_logs_505.txt](https://github.com/TIBCOSoftware/bw6-plugin-maven/files/5384872/maven_logs_505.txt)

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : As i am using the same set of jars as provided, what could have gone wrong? Are you able to make anything out of the debug logs provided?

**Comment by vpawar-ai** [comment]
Please ensure the fixed jars are present in local m2 repo as well as on JFrog.

**Comment by bibhuduttaGitHub** [comment]
Yes, already verified that. Deleted .m2/repository as well to ensure it downloads rhe jar from JFrog. But still no clue why null pointer is coming. PLease suggest. The log attached last time still valid and latest.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : I figured out the jars which were causing the null pointer exceptions. however fixing those jars , it was able to build successfully but post deployment, it is giving below startup errors. I also tried to deploy studio generated ear and it was able to start without issues. please check.

ERROR [pool-7-thread-1] c.t.b.thor.runtime.model.Constituent - The following error has occurred for "name: EDWAdapter version: 1.0.18.202010200947 bundle name: EDWAdapter " which needs to be resolved.
14:04:22.106 ERROR [pool-7-thread-1] c.t.b.thor.runtime.model.Constituent - TIBCO-BW-FRWK-000001: parsing composite
14:04:22.633 ERROR [pool-7-thread-1] com.tibco.thor.frwk.Application - TIBCO-THOR-FRWK-600026: Failed to start BW Application [EDWAdapter.application:1.0]. Reason: TIBCO-BW-FRWK-000001: parsing composite.
   <CausedBy> java.lang.ClassCastException: org.apache.xerces.parsers.IntegratedParserConfiguration cannot be cast to org.apache.xerces.xni.parser.XMLParserConfiguration

**Comment by vpawar-ai** [comment]
@bibhuduttaGitHub Based on the exception, I suspect there is an issue of transitive dependencies added for custom java code in the application. Those dependencies \ libs are conflicting with existing libs and multiple Xerces implementations are found in the classpath. Please review all dependencies and add conflicting dependencies in exclusion list. 
Please refer - https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html#dependency-exclusions

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub, Please check if `stax2-api` and `woodstox-core-lgpl` are required. I removed these dependencies from `EDWAdapter/pom.xml` file.
```xml
<dependency>
    <groupId>stax2-api</groupId>
    <artifactId>stax2-api</artifactId>
    <version>3.0.1</version>
</dependency>
<dependency>
    <groupId>woodstox-core-lgpl</groupId>
    <artifactId>woodstox-core-lgpl</artifactId>
    <version>4.0.4</version>
</dependency>
```
Also please check `EDWAdapter\MANIFEST.MF`, the `Bundle-ClassPath` value should be -
```
Bundle-ClassPath: .
```
After making these changes, I could run the application without any issues. Attaching logs for your reference.
[logs_505_working.txt](https://github.com/TIBCOSoftware/bw6-plugin-maven/files/5451014/logs_505_working.txt)

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : I tried and i still get the same classcast exception.

**Comment by vpawar-ai** [comment]
@bibhuduttaGitHub, The application attached to the case just had below list dependencies, so the EAR contains below jars -

![image](https://user-images.githubusercontent.com/29373785/97427399-749a5100-193a-11eb-9651-4fa45cb31945.png)

pom.xml -
```xml
 <dependencies>
	<dependency>
		<groupId>com.tibco.plugins</groupId>
		<artifactId>com.tibco.bw.palette.shared</artifactId>
		<version>6.1.100</version>
		<scope>provided</scope>
	</dependency>
	<dependency>
		<groupId>edw</groupId>
		<artifactId>edw</artifactId>
		<version>edw</version>
	</dependency>
	<dependency>
		<groupId>junit</groupId>
		<artifactId>junit</artifactId>
		<version>3.8.1</version>
	</dependency>
	<dependency>
		<groupId>log4j</groupId>
		<artifactId>log4j</artifactId>
		<version>1.2.15</version>
		<exclusions>
			<exclusion>
				<groupId>javax.jms</groupId>
				<artifactId>jms</artifactId>
			</exclusion>
			<exclusion>
				<groupId>com.sun.jdmk</groupId>
				<artifactId>jmxtools</artifactId>
			</exclusion>
			<exclusion>
				<groupId>com.sun.jmx</groupId>
				<artifactId>jmxri</artifactId>
			</exclusion>
		</exclusions>
	</dependency>
	<dependency>
		<groupId>mail</groupId>
		<artifactId>mail</artifactId>
		<version>1.4</version>
	</dependency>
	<dependency>
		<groupId>emsit</groupId>
		<artifactId>qct_tibco_common</artifactId>
		<version>1.1.3</version>
	</dependency>
	<dependency>
		<groupId>emsit</groupId>
		<artifactId>qct_util</artifactId>
		<version>1.2.46</version>
	</dependency>
	<!-- dependency>
    <groupId>stax2-api</groupId>
    <artifactId>stax2-api</artifactId>
    <version>3.0.1</version>
</dependency>
<dependency>
    <groupId>woodstox-core-lgpl</groupId>
    <artifactId>woodstox-core-lgpl</artifactId>
    <version>4.0.4</version>
</dependency -->
	<dependency>
		<groupId>jsr173_1.0_api</groupId>
		<artifactId>jsr173_1.0_api</artifactId>
		<version>jsr173_1.0_api</version>
	</dependency>
</dependencies>
```
Please review the list of dependencies according to your custom java code. I see `xercesImpl-2.2.1.jar` in the list.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : I tried to peep into the code, but i don't see any dependency on xercesImpl-2.2.1.jar. Not sure, why its getting downloaded during maven build stage. I verified the JAVA_HOME settings and its pointing to  1.8. Even when i have the dependency for the  7 jars in the POM file, i see around 30 odd jars are getting downloaded and packaged in the ear. xercesImpl-2.2.1.jar is one of them. Can you suggest here?

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub, Please execute `mvn dependency:tree -Dverbose` on "EDWAdapter.application.parent" pom.xml to understand hierarchy of the dependencies.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : attached is the POM.xml which helped me to resolve the classcast exception.
Also , can you check why i am seeing  multiple transitive dependent jars packaged in the ear. I hope this is not the case for you. Thats why you never encountered this classcast exception.
[pom_rename_to_dot_xml.txt](https://github.com/TIBCOSoftware/bw6-plugin-maven/files/5486821/pom_rename_to_dot_xml.txt)
You can see the details of the transitive  jars which are getting downloaded  in the below link:
https://github.com/TIBCOSoftware/bw6-plugin-maven/issues/505#issuecomment-717850268

**Comment by vpawar-ai** [comment]
Hi @bibhuduttaGitHub, I see a difference. On my machine, dependency jars were installed into local m2 repo using `mvn install:install-file` command which created a basic pom.xml file without any dependencies.
When I replaced it with pom.xml from inside the jar, I could see transitive dependencies jars getting added into the EAR. As discussed during the call, using maven exclusions list is the way forward and it fixes the issue for you.  Please reopen the case or create new one if you need any additional help.  Thanks.

**Comment by bibhuduttaGitHub** [comment]
@vpawar-ai : Most of our projects got resolved after applying required exclusion and scope for the jars.
However i see null pointer exception for a project, i have already validated the jar packaging of the dependent jars and transitive jars. I don't see any issue with respect to jar packaging. But still i get the null pointer exception. I have attached the project(Novatus) and the build debug logs and customxpathfunction jars to the case: 01888935. Take a look at them.
